contracts/modules/Experimental/Checkpoints/WeightedVoteCheckpoint.sol

```diff
@@ -0,0 +1,188 @@
+        require(ballot.totalProposals >= _proposalId && _proposalId > 0, "Incorrect proposals Id");
+        require(now >= ballot.startTime && now <= ballot.endTime, "Voting period is not active");
+        require(ballot.voteByAddress[msg.sender].voteTime == 0, "Token holder has already voted");
+        require(ballot.isActive == true, "Ballot is not active");
+
+        
+        ballot.voteByAddress[msg.sender].voteTime = now;
+        ballot.voteByAddress[msg.sender].weight = weight;
+        ballot.voteByAddress[msg.sender].vote = _proposalId;
+        ballot.totalNumVotes = ballot.totalNumVotes.add(1);
+        ballot.proposalToVote[_proposalId] = ballot.proposalToVote[_proposalId].add(weight);
+        emit VoteCasted(_ballotId, _proposalId, msg.sender, weight);
+    }
+
+    /**
+     * @notice Allows the token issuer to set the active stats of a ballot
+     * @param _ballotId The index of the target ballot
+     * @param _isActive The bool value of the active stats of the ballot
+     * @return bool success
+     */
+    function changeBallotStatus(uint256 _ballotId, bool _isActive) external {
+        _onlySecurityTokenOwner();
+        require(now < ballots[_ballotId].endTime, "Already ended");
+        require(ballots[_ballotId].isActive != _isActive, "Active state unchanged");
+        ballots[_ballotId].isActive = _isActive;
+        emit BallotStatusChanged(_ballotId, _isActive);
+    }
+
+    /**
+     * @notice Queries the result of a given ballot
+     * @param _ballotId Id of the target ballot
+     * @return uint256 abstain
+     * @return uint256 remainingTime
+     * @return uint256 totalVotes
+     */
+    function getResults(uint256 _ballotId) external view returns (
+        uint256[] memory,
+        uint256 remainingTime,
+        uint256 totalVotes
+    ) {
+        if (_ballotId < ballots.length) {
+            Ballot storage ballot = ballots[_ballotId];
+            uint256[] memory abstain = new uint256[](ballot.totalProposals);
+            for (uint256 i = 0; i < ballot.totalProposals; i++) {
+                abstain[i] = ballot.proposalToVote[i+1];
+            }
+            if (ballot.endTime >= now)
+                remainingTime = ballot.endTime.sub(now);
+            totalVotes = ballot.totalNumVotes;
+            return (abstain, remainingTime, totalVotes);
+        }
+        return (new uint256[](0), remainingTime, totalVotes);
+    }
+
+    /**
+     * @notice Get the stats of the ballot
+     * @param _ballotId The index of the target ballot
+     */
+    function getBallotStats(uint256 _ballotId) external view returns(uint256, uint256, uint256, uint256, uint256, uint256, bool) {
+        Ballot memory ballot = ballots[_ballotId];
+        return (
+            ballot.checkpointId,
+            ballot.totalSupply,
+            ballot.startTime,
+            ballot.endTime,
+            ballot.totalNumVotes,
+            ballot.totalProposals,
+            ballot.isActive
+        );
+    }
+
```

Comment:
I think we should add a getter to return the proposal selected by the voter. 
It should take ballot id and voter address as input and return the proposal selected.